### PR TITLE
Add macOS exporting and running pages.

### LINKED ISSUE #5076: Export to macOS: Document scenarios and options for devs and players regarding code signing and notarization
**Your Godot version:**
3.3 and later

**Issue description:**
Exporting a project for macOS has lots of caveats regarding Apple requirements on code signing, notarization and entitlements, especially on M1 macs.

We need to document this thoroughly in order to:
- Let developers know what they should be aware of when exporting for macOS
- Let players know how they can play a Godot-made app if it was not fully signed and notarized by the developer (or more likely, let the developers know what they should tell users to do)

Some input from @bruvzg:

> Currently, there are three possible scenarios:
> 1. Signed and notarized app - app just run.
> 2. Ad-hoc signed app (default behavior on export from macOS) - won't run, but uses can override it by right-clicking app in the "Finder" selecting "Open" and confirming (should be done twice), or running `xattr` command.
> 3. Not signed app (with the linker-signed executable) - won't run, the only way to override is running `xattr -dr com.apple.quarantine AppName.app` command. If the executable is not linker-signed, it won't run on Apple Silicon macs at all.

The document should also explain the differences between exporting for macOS *from* macOS, and from other platforms (Linux and Windows) where code signing cannot be performed by the editor (and thus needs to be done manually on a Mac - or it's point 3. above).

**URL to the documentation page (if already existing):**
https://docs.godotengine.org/en/stable/getting_started/workflow/export/exporting_for_pc.html

That page should likely be split into 3 separate pages for Linux, Windows and macOS so that we can document platform-specific quirks properly.

### PR DESCRIPTION
**Pull request by bruvzg**
TODO:
- [x] Running guide for exported app users.
- [x] Debugging Godot editor on macOS. 
- [ ] Add "Export for macOS" tutorial.
   - [x] Singing and notarization
   - [x] Entitlements.
- [ ] Split "Export for PC" to Windows/Linux pages.
   - [x] Export for Linux page
   - [x] Export for Windows page
   - [ ] Code signing on Windows.

Fixes #5076, #5066, #4927, #2739

### CODE REVIEW

**Comment by ee0pdt** [comment]
Singing is optional btw… but it helps maintain morale 😉

File: tutorials/export/exporting_for_mac.rst [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+You can use ``xcrun notarytool history`` command to check notarization status, and use ``xcrun notarytool log {ID}`` command to download notarization log.
+
+If you encounter notarization issues, see `Resolving common notarization issues <https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution/resolving_common_notarization_issues>`__ for more info.
+
+After notarization is completed, `staple the ticket <https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution/customizing_the_notarization_workflow>`__ to the exported project.
+
+If you do not have an Apple Developer ID Certificate
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Keep "Code Signing" enabled and leave "identinty" option empty, in this case Godot will use ad-hoc signature, which will make running an exported app easier for the end users.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:43]:
```suggestion
Keep ``Code Signing`` enabled and leave the ``Identity`` option empty, in this case Godot will use a ad-hoc signature, which will make running an exported app easier for the end users.
```

File: tutorials/export/exporting_for_mac.rst [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+| Timestamp                    | Requests a timestamp server to authenticate the time of signing. Required for notarization.       |
++------------------------------+---------------------------------------------------------------------------------------------------+
+| Hardened Runtime             | Enables "Hardened Runtime". Required for notarization.                                            |
++------------------------------+---------------------------------------------------------------------------------------------------+
+| Replace Existing Signature   | Replaces existing signatures of the GDNative libraries and embedded helper executables.           |
++------------------------------+---------------------------------------------------------------------------------------------------+
+
+.. note::
+
+    To notarize an app, you must enable the "Hardened Runtime" and "Timestamp".
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:64]:
```suggestion
    To notarize an app, you must enable the ``Hardened Runtime`` and ``Timestamp``.
```

File: tutorials/export/exporting_for_mac.rst [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+.. [1] Leave "identinty" option empty to use ad-hoc signature.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:66]:
```suggestion
.. [1] Leave ``Identity`` option empty to use ad-hoc signature.
```

File: tutorials/export/exporting_for_mac.rst [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+| Apple Events                          | [3]_ Enable to allow app to send Apple events to other apps.                                                                                                                                     |
++---------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Debugging                             | [4]_ You can temporarily enable this entitlement to use native debugger (GDB, LLDB) with the exported app. This entitlement should be disabled for production export.                            |
++---------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+
+.. [2] "Allow JIT Code Execution", "Allow Unsigned Executable Memory" and "Allow DYLD Environment Variables" entitlements are always enabled for the Godot Mono exports, and are not visible in the export options.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:122]:
```suggestion
.. [2] The ``Allow JIT Code Execution``, ``Allow Unsigned Executable Memory`` and ``Allow DYLD Environment Variables`` entitlements are always enabled for the Godot Mono exports, and are not visible in the export options.
```

File: tutorials/export/exporting_for_mac.rst [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+.. [3] These features aren't supported by Godot out of the box, enable it only if you are using add-ons which require it.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:123]:
```suggestion
.. [3] These features aren't supported by Godot out of the box, enable them only if you are using add-ons which require them.
```

File: tutorials/export/exporting_for_mac.rst [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+.. [4] To notarize an app, you must disable "Debugging" entitlement.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:124]:
```suggestion
.. [4] To notarize an app, you must disable the ``Debugging`` entitlement.
```

File: tutorials/export/exporting_for_mac.rst [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+App Sandbox Entitlement
+~~~~~~~~~~~~~~~~~~~~~~~
+
+App Sandbox restrict access to user's data, network and devices.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:129]:
```suggestion
The App Sandbox restricts access to user data, networking and devices.
```

File: tutorials/export/exporting_for_mac.rst [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+Sandboxed apps can't access most of the file system, can't use custom file dialogs and execute binaries (using ``OS.execute`` and ``OS.create_process``) outside the ``.app`` bundle.
+See `App Sandbox <https://developer.apple.com/documentation/security/app_sandbox?language=objc>`__ for more info.
+
+.. note::
+
+    To distribute an app through the App Store, you must enable App Sandbox.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_mac.rst:135]:
```suggestion
    To distribute an app through the App Store, you must enable the App Sandbox.
```

File: tutorials/export/exporting_for_windows.rst [hunk 9]
```diff
@@ -0,0 +1,29 @@
+.. _doc_exporting_for_windows:
+
+Exporting for Windows
+=====================
+
+The simplest way to distribute a game for PC is to copy the executables
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_windows.rst:6]:
```suggestion
The simplest way to distribute a game for PC is to copy the executable
```

File: tutorials/export/exporting_for_windows.rst [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+Godot offers a more elegant approach for PC distribution when using the export
+system. When exporting for Windows, the exporter takes all the project files and
+creates a ``data.pck`` file. This file is bundled with a specially optimized
+binary that is smaller, faster and does not contain the editor and debugger.
+
+Requirements
+------------
+
+-  To enable code signing, you must have ``Windows 10 SDK`` (on Windows) or `osslsigncode <https://github.com/mtrojnar/osslsigncode>`__ (on any other OS) installed.
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_windows.rst:18]:
```suggestion
-  To enable code signing, you must have the ``Windows 10 SDK`` (on Windows) or `osslsigncode <https://github.com/mtrojnar/osslsigncode>`__ (on any other OS) installed.
```

File: tutorials/export/exporting_for_windows.rst [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+-  Download the Godot export templates. Use the Godot menu: Editor > Manage Export Templates
```
Comment (mhilbrunner) [inline tutorials/export/exporting_for_windows.rst:19]:
```suggestion
-  Download the Godot export templates. Use the Godot menu: ``Editor > Manage Export Templates``.
```

File: tutorials/export/running_on_mac.rst [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+By default, macOS will run only applications that are signed and notarized.
+
+Depending on the way a macOS app is signed and distributed, the following scenarios are possible:
+
+App is signed, notarized and distributed via App Store
+------------------------------------------------------
+
+.. note::
+
+    App developer have joined Apple Developer Program, and configured signing and notarization options during export, and uploaded it to the App Store.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:15]:
```suggestion
    App developers need to join the Apple Developer Program, and configure signing and notarization options during export, then upload the app to the App Store.
```

File: tutorials/export/running_on_mac.rst [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+The app should run out of the box, without extra user interaction required.
+
+App is signed, notarized and distributed outside App Store
+----------------------------------------------------------
+
+.. note::
+
+    App developer have joined Apple Developer Program, and configured signing and notarization options during export, and distributed as ".DMG" or ".ZIP" archive.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:24]:
```suggestion
    App developers need to join the Apple Developer Program, and configure signing and notarization options during export, then distribute the app as ".DMG" or ".ZIP" archive.
```

File: tutorials/export/running_on_mac.rst [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+When you run the app for the first time, the following dialog is displayed:
+
+.. image:: img/signed_and_notarized_0.png
+
+Click "Open" to start the app.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:30]:
```suggestion
Click ``Open`` to start the app.
```

File: tutorials/export/running_on_mac.rst [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+If you see the following warning dialog, your Mac is set up to allow apps only from the App Store.
+
+.. image:: img/signed_and_notarized_1.png
+
+To allow third-party apps open "System Preferences", click "Security & Privacy", then click "General", unlock settings, and select "App Store and identified developers".
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:36]:
```suggestion
To allow third-party apps, open ``System Preferences``, click ``Security & Privacy``, then click ``General``, unlock settings, and select ``App Store and identified developers``.
```

File: tutorials/export/running_on_mac.rst [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+    App developer used self-signed certificate or ad-hoc signing (default Godot behavior for exported project).
+
+When you run the app for the first time, the following dialog is displayed:
+
+.. image:: img/signed_0.png
+
+To run this app, you can temporarily override Gatekeeper:
+
+* Either open "System Preferences", click "Security & Privacy", then click "General", and click "Open Anyway".
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:53]:
```suggestion
* Either open ``System Preferences``, click ``Security & Privacy``, then click ``General``, and click ``Open Anyway``.
```

File: tutorials/export/running_on_mac.rst [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+  .. image:: img/sys_pref_1.png
+
+* Or, right-click (Control-click) on the app icon in the "Finder" window and select "Open" from the menu.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:57]:
```suggestion
* Or, right-click (Control-click) on the app icon in the Finder window and select ``Open`` from the menu.
```

File: tutorials/export/running_on_mac.rst [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+  .. image:: img/signed_1.png
+
+* Then click "Open" in the confirmation dialog.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:61]:
```suggestion
* Then click ``Open`` in the confirmation dialog.
```

File: tutorials/export/running_on_mac.rst [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+    App is built using official export templates, but it is not signed.
+
+When you run the app for the first time, the following dialog is displayed:
+
+.. image:: img/linker_signed_1.png
+
+To run this app, you should remove "quarantine" extended file attribute manually:
+
+* Open "Terminal.app" (press ``Cmd + Space``, and enter "Terminal").
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:80]:
```suggestion
* Open ``Terminal.app`` (press ``Cmd + Space``, and enter ``Terminal``).
```

File: tutorials/export/running_on_mac.rst [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+.. note::
+
+    App is built using official export templates, but it is not signed.
+
+When you run the app for the first time, the following dialog is displayed:
+
+.. image:: img/linker_signed_1.png
+
+To run this app, you should remove "quarantine" extended file attribute manually:
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:78]:
```suggestion
To run this app, you should remove the quarantine extended file attribute manually:
```

File: tutorials/export/running_on_mac.rst [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+* Open "Terminal.app" (press ``Cmd + Space``, and enter "Terminal").
+
+* Navigate to the folder containing the target application.
+
+  Use ``cd path_to_the_app_folder`` command, e.g. ``cd ~/Downloads/`` if it's in the "Downloads" folder.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:84]:
```suggestion
  Use the ``cd path_to_the_app_folder`` command, e.g. ``cd ~/Downloads/`` if it's in the ``Downloads`` folder.
```

File: tutorials/export/running_on_mac.rst [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+
+* Run following command ``xattr -dr com.apple.quarantine "Unsigned Game.app"`` (including quotation marks and ".app" extension).
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:86]:
```suggestion
* Run the command ``xattr -dr com.apple.quarantine "Unsigned Game.app"`` (including quotation marks and ``.app`` extension).
```

File: tutorials/export/running_on_mac.rst [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+
+* Install ``Xcode`` for the App Store, start it and confirm command line tools installation.
+
+* Open "Terminal.app" (press ``Cmd + Space``, and enter "Terminal").
+
+* Navigate to the folder containing the target application.
+
+  Use ``cd path_to_the_app_folder`` command, e.g. ``cd ~/Downloads/`` if it's in the "Downloads" folder.
+
+* Run following commands:
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:109]:
```suggestion
* Run the following commands:
```

File: tutorials/export/running_on_mac.rst [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+
+When you run the app for the first time, the following dialog is displayed:
+
+.. image:: img/unsigned_1.png
+
+To run this app, you can ad-hoc sign it yourself:
+
+* Install ``Xcode`` for the App Store, start it and confirm command line tools installation.
+
+* Open "Terminal.app" (press ``Cmd + Space``, and enter "Terminal").
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:103]:
```suggestion
* Open ``Terminal.app`` (press ``Cmd + Space``, and enter ``Terminal``).
```

File: tutorials/export/running_on_mac.rst [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+
+* Navigate to the folder containing the target application.
+
+  Use ``cd path_to_the_app_folder`` command, e.g. ``cd ~/Downloads/`` if it's in the "Downloads" folder.
```
Comment (mhilbrunner) [inline tutorials/export/running_on_mac.rst:107]:
```suggestion
  Use ``cd path_to_the_app_folder`` command, e.g. ``cd ~/Downloads/`` if it's in the ``Downloads`` folder.
```

**Review by mhilbrunner — COMMENTED** [review]
Looks great :) Is there anything left to do?

I left some nitpicks - sorry, these look like a lot, but 90% of it is changing a lot of quotes to `` and fixing articles.

If you want, I can apply these style edits myself, as I figure you're pretty busy. :)

**Comment by bruvzg** [comment]
I have applied all suggested changes.

> Is there anything left to do?

No, everything macOS seems to be done.

**Comment by mhilbrunner** [comment]
Wow, amazing work - and that was fast :) Thanks a ton!

File: development/cpp/macos_debug.rst [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+            <key>com.apple.security.device.audio-input</key>
+            <true/>
+            <key>com.apple.security.device.camera</key>
+            <true/>
+            <key>com.apple.security.cs.disable-library-validation</key>
+            <true/>
+            <key>com.apple.security.get-task-allow</key>
+            <true/>
+        </dict>
+    </plist>
```
[the remark this answers is not in the document]
Comment (bruvzg) [reply development/cpp/macos_debug.rst:27]: Yes, should be full set of entitlements from #54067 plus:
```xml
<key>com.apple.security.get-task-allow</key>
<true/>
```

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Pull request merged**